statem_clnt: do not override version on sess reuse

indutny opened the pull request:
Allow downgrading when reusing sessions on client

When connecting to pool of diverse servers (both TLS1.0 and TLS1.2), a
following scenario may happen:

  1. Connect to TLS1.2 server, receive new session
  2. Store this session
  3. Attempt to reuse it later when connecting to server
  4. Connect to different server from the pool, which speaks only TLS1.0
  5. Get `SSL_R_WRONG_VERSION_NUMBER` error

Expected behavior would be scrapping off the session, and allowing
server to downgrade to supported protocol version the way it would do it
if no client session would be supplied.

cc @mattcaswell 

Supersedes #603

kroeckx commented:
So if I understand this correctly, one of the changes is that even though we expect to reuse the session with TLS 1.0, we're going to send a client hello using TLS 1.2 (assuming TLS 1.2 is enabled)?

kroeckx commented:
As far as I can see, the *_get_method functions are currently only used in SSL_set_session() now, so that change looks safe to do in master, but I doubt that's good for the stable branches.

I'm wondering if we shouldn't simplify the SSL_set_session() code to just use s->ctx->method, and remove all those *_get_method() functions.

In any case, you didn't seem to have made the same change for the DTLS functions.

kroeckx commented on test/heartbeat_test.d.tmp:
```diff
@@ -0,0 +1,18 @@
+test/heartbeat_test.o: test/heartbeat_test.c test/../ssl/ssl_locl.h \
+  e_os.h include/openssl/opensslconf.h include/openssl/e_os2.h \
```
I think you shouldn't add this file.

indutny replied:
Ack.

kroeckx commented on ssl/statem/statem_lib.c:
```diff
@@ -969,14 +969,6 @@ int ssl_choose_client_version(SSL *s, int version)
-         * that's OK.  It is up to the caller to not choose fixed protocol
-         * versions they don't want.  If not, then easy to fix, just return
-         * ssl_method_error(s, s->method)
-         */
```
You probably shouldn't remove that comment, it has nothing to do with your change and is still correct.

indutny replied:
But the relevant code is gone in this patch. Should I move it to client state machine file then?

indutny replied:
Moved.

kroeckx replied:
I'm not sure what you mean.  The case "default:" case is for when a version-specific method is selected.  It's saying that it's not doing the same checks as for the version-flexible methods to see if this protocol should be enabled or not, it's always enabled.  The comment is not about the next line.

I'm also unsure what you mean with "moved".

indutny replied:
Ack.

kroeckx commented on ssl/statem/statem_lib.c:
```diff
@@ -1007,6 +999,34 @@ int ssl_choose_client_version(SSL *s, int version)
+ *
+ * @s: client SSL handle.
+ *
+ * Returns 1 on success, 0 on failure
```
Can you say something like "1 when it matches, 0 otherwise"?

indutny replied:
Ack.

kroeckx commented on ssl/statem/statem_lib.c:
```diff
@@ -1007,6 +999,34 @@ int ssl_choose_client_version(SSL *s, int version)
+  }
+
+  /* Empty session and no EAP-FAST */
+  if (s->session->session_id_length == 0 && !s->session->tlsext_tick) {
```
Please use `s->session->tlsext_tick == NULL`

kroeckx replied:
Should that be `||` instead of `&&`?

indutny replied:
As far as I understood it `tlsext_tick` is non-`NULL` when we are doing `EAP-FAST`, and `session_id` is empty. This condition attempts to catch empty sessions that we just created by the state machine itself.

kroeckx commented on ssl/statem/statem_clnt.c:
```diff
@@ -1136,6 +1136,21 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
                session_id_len);
     }
 
+    if (!ssl_verify_client_session_version(s)) {
```
So I'm confused about what it's trying to do.  I assume that this is the case of session resumption, and so `s->hit == 1`?  Are you trying to detect session resumption in some other way, and then check that the version is the one you expected?

kroeckx replied:
So below we have have this code:

    if (s->hit && (s->session->cipher_id != c->id)) {
         al = SSL_AD_ILLEGAL_PARAMETER;
         SSLerr(SSL_F_TLS_PROCESS_SERVER_HELLO,
                SSL_R_OLD_SESSION_CIPHER_NOT_RETURNED);
         goto f_err;
     }

Maybe we should so do something like that here?

davidben replied:
This is what BoringSSL does, along with every other client I've tested. If the version doesn't match, it's a fatal error. Switching to a fresh session isn't an option anyway. `s->hit` is still 1, so we'll try to resume an empty master secret. (And setting `s->hit` back to 0 will cause the handshake to break. The server will continue the handshake as if it's a resumption handshake.)

indutny replied:
Yeah, `s->hit` is not a guarantee here. I think this is the most safe way to do it atm.

davidben replied:
So, this version isn't actually safe, I think. If you call `ssl_get_new_session` and leave `s->hit` at 1, you'll resume an empty master secret and be trivially MITM-able.

indutny replied:
This is a error case... Does `s->hit` make difference when we are killing connection?

kroeckx replied:
So why do you get a new session?

davidben replied:
Oh, I see. I think it's best to avoid calling `ssl_get_new_session` then. It doesn't actually do anything. Having `s->hit` be 1 and `s->session` blank also violates a pretty important invariant.

indutny replied:
Well, we do get new session in all other error cases... I was just mimicking other places in this file ;)

So I can either set `s->hit = 0;` or remove `ssl_get_new_session` code. Which one would you prefer? (They seem to be the same to me).

kroeckx replied:
@davidben: So 1 thing I'm confused about when you say "this is what BoringSSL does", does it check for s->hit before comparing the versions?

davidben replied:
Which places are you thinking of? The two `ssl_get_new_session` calls I see aren't error cases but cases where we decide not to reuse a session. This one is a straight-up error path, most analogous to the `SSL_R_OLD_SESSION_CIPHER_NOT_RETURNED` case which doesn't reset `s->session`.

indutny replied:
Oh gosh, you are right. They are error cases only when `ssl_get_new_session` fails. Fixed.

davidben replied:
(Ugh, my last comment was in reply to @indutny. GitHub's UI is so bad at concurrent conversations...)

@kroeckx Yeah, I mean that the version match check is made very similarly to the cipher match check in BoringSSL and that it's fatal. (I'd missed that it was fatal in the PR too. The `ssl_get_new_session` call threw me off and I can't read GitHub diffs. :-) )

kroeckx commented:
So I understand that this change will fix the client so it tells the server it supports a higher version, but the server might still want to reuse it the session?  Maybe the server should be changed so that it doesn't reuse the session in case the client offers a newer version than what's in the session?  Or was that part already fixed in master?

davidben commented:
@kroeckx: Yeah, the server behavior was fixed in https://git.openssl.org/gitweb/?p=openssl.git;a=commitdiff;h=03d14f58873470407de6120218b7e69fefd8b58f

indutny commented:
Looks like I will have to `i2d` + `d2i` session in tests.

davidben commented on ssl/statem/statem_clnt.c:
```diff
@@ -1136,6 +1136,21 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
+    }
+
+    /* It is safe to assign version for a new session at this point */
+    s->session->ssl_version = s->version;
```
I think this still isn't quite safe. An `SSL_SESSION` should be completely immutable once established. The problem is another thread may be using the object while this assignment is happening. The write is a no-op, but it's still a race. (It's also unnecessary because of the ssl_verify_client_session_version check.)

indutny replied:
Let me run tests to confirm this.

indutny replied:
Tests pass.

davidben commented on ssl/statem/statem_lib.c:
```diff
@@ -1007,6 +999,34 @@ int ssl_choose_client_version(SSL *s, int version)
+{
+  int err;
+
+  if (s->session == NULL) {
```
I think it's impossible for s->session to be NULL at this point.

indutny replied:
This is what I thought too.

indutny replied:
Ack.

indutny commented:
Pushed fixes. Working on fixing shared tests.

davidben commented on ssl/methods.c:
```diff
@@ -118,26 +118,7 @@
-    else
-#endif
-    return NULL;
+    return TLS_method();
```
Rather than no-op-ing the `get_ssl_method` call in `SSL_set_session`, I think it's better to remove it. This does some odd things for the version-locked methods. (Granted, the old code does odd things too, albeit different ones.)

indutny replied:
I think it may make it more complex to backport this patch. I don't mind doing it in separate PR, though.

kroeckx replied:
I think I made the same comment above?

davidben replied:
I'll defer to OpenSSL, but given how different 1.0.2 and 1.1.0 are, it'll already be a non-trivial backport. I would probably do it as correctly as possible for 1.1.0 and, once that's working, figure out how to do it in 1.0.2. (For instance, you could leave the `get_ssl_method` hook for ABI reasons but remove the call in `SSL_set_session`. The hook would just be unused within the library.)

davidben replied:
@kroeckx Ah, sorry, I must have missed that!

indutny replied:
Removed.

indutny commented:
Hopefully, fixed travis build.

davidben commented on ssl/statem/statem_lib.c:
```diff
@@ -1007,6 +1007,30 @@ int ssl_choose_client_version(SSL *s, int version)
+ */
+int ssl_verify_client_session_version(SSL *s)
+{
+  int err;
```
Nit: unused

indutny replied:
Fixed.

kroeckx commented on ssl/statem/statem_lib.c:
```diff
@@ -1007,6 +1007,28 @@ int ssl_choose_client_version(SSL *s, int version)
+  }
+
+  return 0;
+}
```
So I guess that there is 1 thing I'm still unclear about, and I guess that's about EAP-FAST.  I know it's special, but I always forget the details.

indutny replied:
Yeah, it seems that in this case there is just a TLS session ticket and empty session id.

davidben commented on ssl/ssl_sess.c:
```diff
@@ -877,13 +877,7 @@ int SSL_set_session(SSL *s, SSL_SESSION *session)
+        meth = s->ctx->method;
 
         if (meth != s->method) {
             if (!SSL_set_ssl_method(s, meth))
```
I think you want to lose this (and line 903) too? The only case where it runs is if version negotiation has already happened. This API already explodes horrible if called after the `SSL` has started, so it would only be relevant if you reuse the object with `SSL_clear`. (This is a really bad feature and needs to go away. Alas, I wasn't able to for BoringSSL. Found enough consumers that did it.) But, in that case, the method reset belongs in `SSL_clear`, not in `SSL_set_session`.

indutny replied:
Oh gosh. So many refactoring suggestions :) Perhaps they should come in separate PR?

davidben replied:
Yeah, it really is a never-ending cycle with this part of the code. :-)

kroeckx replied:
So also get rid of the meth variable?

I assume that you only call SSL_set_session() after SSL_new()?

indutny replied:
Ack.

davidben commented on ssl/statem/statem_lib.c:
```diff
@@ -1007,6 +1007,28 @@ int ssl_choose_client_version(SSL *s, int version)
+int ssl_verify_client_session_version(SSL *s)
+{
+  /* Empty session and no EAP-FAST */
+  if (s->session->session_id_length == 0 && !s->session->tlsext_tick) {
```
Oh. This might actually be hiding another bug. Sorry, my bad! I should have realized this when I pointed out the `session->ssl_version = version` line. Because `ssl_get_new_session` is called *before* version resumption in `tls_construct_client_hello`, the `ssl_version` field isn't initialized right.

One option is to remove this check and restore the ssl_version assignment, but in line 1132, where you know the session is fresh. Another option, and what I did for BoringSSL, is to take the `ssl_get_new_session` call out of `tls_construct_client_hello` and uncondition the other one. But that's a more involved change. (It means `s->session` can be NULL in more states.)

I think you'd be able to reproduce this with tests that ensure session resumption happens at all versions. https://rt.openssl.org/Ticket/Display.html?id=4392 is an example of what goes wrong. (OpenSSL currently has a bug where `ssl_version` isn't initialized right in DTLS on the server half.)

(Actually, you might want to play with https://mta.openssl.org/pipermail/openssl-dev/2016-March/005779.html. The bug you're fixing here is caught by BoringSSL tests.)

NB: I also have no clue about EAP-FAST and how this interacts. We've removed that code because it was crazy. :-)

indutny replied:
Took some time to figure it out, but I think I got it. Put the `session->ssl_version = ` assignment back, but not sure which check you think should be removed.

I don't really want to go into more involved approach, but again it may be a good idea to do this after this PR. Right now, I'm trying to fix a bug that node.js users are experiencing.

davidben replied:
> Took some time to figure it out, but I think I got it. Put the session->ssl_version = assignment back,

Oh, I was thinking it should go slightly higher, near `s->session->session_id_length = session_id_len;`. This way you fix `ssl_version` in the fresh session case, but don't violate thread-safety in reuse.

> but not sure which check you think should be removed.

The one this thread is attached to. `if (s->session->session_id_length == 0 && !s->session->tlsext_tick) {`. Or is that for something else? I think this function really just wants to be a version check.

> I don't really want to go into more involved approach, but again it may be a good idea to do this after this PR. Right now, I'm trying to fix a bug that node.js users are experiencing.

Sure. I did a minimal fix before the more aggressive one myself. :-)

indutny replied:
>Oh, I was thinking it should go slightly higher, near s->session->session_id_length = session_id_len;. This way you fix ssl_version in the fresh session case, but don't violate thread-safety in reuse.

@davidben like this? Added assignment in two places.

>The one this thread is attached to. if (s->session->session_id_length == 0 && !s->session->tlsext_tick) {. Or is that for something else? I think this function really just wants to be a version check.

@davidben removed the branch, and inlined the function.

kroeckx commented on ssl/statem/statem_clnt.c:
```diff
@@ -1136,6 +1138,24 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
+     * that's OK.  It is up to the caller to not choose fixed protocol
+     * versions they don't want.  If not, then easy to fix, just return
+     * ssl_method_error(s, s->method)
+     */
```
That comment clearly shouldn't be here.

indutny replied:
Ack.

indutny replied:
Fixed, sorry.

kroeckx commented on ssl/statem/statem_clnt.c:
```diff
@@ -1136,6 +1138,24 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
+    /* Session version and negotiated protocol version should match
+     * (This may not be the case when reusing old session)
+     */
+    if (s->version != s->session->ssl_version) {
```
So I think that should be `if (s->hit && s->version != s->session->ssl_version)`

indutny replied:
Well, if `s->hit == 0`, `s->version` should be `s->session->ssl_version` anyway, right? Why bother with extra checks?

mattcaswell replied:
I don't understand why this check is necessary (or the setting of `s->session->ssl_verion = s->version` a few lines above). Isn't this all covered by the `ssl_choose_client_version()` call at the beginning of this function?

mattcaswell replied:
Never mind....I see you've changed ssl_choose_client_version() below.

mattcaswell replied:
So now I don't understand the comment! :-) If we attempted to reuse an old session, but failed, then `s->hit == 0` and `s->version == s->session->ssl_version`? So how can this ever fail except through server error?

indutny replied:
This handles the case of `s->hit == 1` with stale session.

mattcaswell replied:
But if its a stale session then `s->hit == 0` no?

indutny replied:
Nope, `if` above will set it to `1`. We are kind of accounting for a faulty server here, that will accept session and negotiate higher protocol version than supported by the session.

mattcaswell replied:
The `if` above only sets it to 1 in the case of a session id match, i.e. a non-stale session?? The faulty server thing I get. That is, I'm not disputing the need for this check...only the comment that describes it (I'd suggest just deleting the second line of the comment).

indutny replied:
Yeah, looks like a terminology clash. I will remove it.

indutny replied:
Ack too.

davidben commented on ssl/statem/statem_clnt.c:
```diff
@@ -830,6 +830,7 @@ int tls_construct_client_hello(SSL *s)
         (sess->not_resumable)) {
         if (!ssl_get_new_session(s, 0))
             goto err;
+        s->session->ssl_version = s->version;
```
This doesn't do anything. `ssl_get_new_session` always initializes it to `s->version`.

indutny replied:
Ack.

davidben commented on ssl/statem/statem_clnt.c:
```diff
@@ -1128,6 +1129,7 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
             if (!ssl_get_new_session(s, 0)) {
                 goto f_err;
             }
+            s->session->ssl_version = s->version;
```
One line down. :-)

The thing you want is for the fresh session to have the same version as the established version. The problem is that, if `session_id_length` is 0---that is, you weren't offering a session at all---we don't call `ssl_get_new_session` here. That means `s->session` is initialized from line 831. But that's before version negotiation!

Thus, the original code "fixed" `ssl_session` afterwards. Because in not all cases was `ssl_session` correct. And this is why I think it's better---but more involved, so it's not the minimal fix---to defer line 831 to here. That way `s->session` has the right version to begin with, rather than the wrong one and then a fixup.

Is that clearer? I feel like I haven't been explaining the issue very well.

indutny replied:
> That means s->session is initialized from line 831. But that's before version negotiation!

This could also mean EAP-FAST case, and I don't want to allow stale session there. (stale = diff. protocol version)

> I feel like I haven't been explaining the issue very well, so let me know if it isn't. The code is pretty gnarly. :-)

Yeah, I guess I figured it out. Sorry for my stupidness! Added an additional check for EAP-FAST case, though.

indutny commented:
Appears to be broken after recent changes... Working on reproducing it locally.

indutny commented:
Travis is green now.

davidben commented:
[Commenting separately because GitHub apparently can't comment on context.]

I just tried running this against BoringSSL's tests and this causes a ton more tests to fail. It looks like there is one more issue here. At the top of `tls_construct_client_hello`, there is some code:

```
    if ((sess == NULL) || (sess->ssl_version != s->version) ||                         
        /*
         * In the case of EAP-FAST, we can have a pre-shared                           
         * "ticket" without a session ID.                                              
         */
        (!sess->session_id_length && !sess->tlsext_tick) ||
        (sess->not_resumable)) {                                                       
        if (!ssl_get_new_session(s, 0))
            goto err;
    }  
```

The `(sess->ssl_version != s->version)` was fine before, because for the version-locking. But now it doesn't work because, at this point, `s->version` is the *maximum* client version, not the final one.

It should be `sess->ssl_version > s->version`, but that doesn't quite work because DTLS invert the comparison because they bit-flip all the versions on the wire (sigh). It looks like `statem_lib.c` has a `version_cmp` helper.

But if this passed tests, probably OpenSSL needs invest in tests before doing any more work on version negotiation. It's really subtle to get right. The BoringSSL test suite I sent to openssl-dev@ might be useful for inspiration.

indutny commented:
@davidben good catch! Working on a fix, and will probably create a test for it.

indutny commented:
@davidben pushed fix, may I ask you to take a look before I will create a test? (or... perhaps, you could point me to the particular BoringSSL tests that revealed this problem?)

davidben commented:
The failed tests were actually just our bog-standard cipher suite tests.
https://boringssl.googlesource.com/boringssl/+/master/ssl/test/runner/runner.go#2218

They run through every version x cipher x {client, server} x {tls, dtls} and do a pair of handshakes. They each fail if the second handshake doesn't resume. That check failed for version < 1.2.

With your fix, they all pass now! :-)

davidben commented on ssl/statem/statem_clnt.c:
```diff
@@ -1130,12 +1133,27 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
         }
 
+        /* Fix session version if it is not reused and we are not in EAP-FAST */
+        if (s->session->tlsext_tick == NULL)
```
Are you sure you don't want this to run in EAP-FAST too? I think the old line 979 would run everywhere. Then you don't need this slightly confusing check.

indutny replied:
How would I reject incorrect `SSL_SESSION`s in EAP-FAST case without this check?

davidben replied:
Hrm. I don't think there are incorrect-version `SSL_SESSION`s to reject in EAP-FAST, actually.

EAP-FAST is basically a messed up version of session tickets, except that you get the ticket and session material from some external mechanism. (That and rather than checking if the server echoed the session ID in ServerHello, you're required to peek ahead for ChangeCipherSpec, which is obnoxious. EAP-FAST doesn't even resolve session resumption until later. See https://github.com/openssl/openssl/blob/master/ssl/statem/statem_clnt.c#L260 But I digress...)

Despite being basically session tickets, OpenSSL's EAP-FAST code isn't structured like this at all. They always make a fresh `SSL_SESSION` and then scribble into the master secret in a callback. See the `tls_session_secret_cb` call. The `ssl_version` isn't filled in from anything, so there's nothing to check. (In fact, if the EAP-FAST consumer even called `SSL_set_session`, the callback would scribble into the master secret which can't possibly work well.)

davidben replied:
(I've just checked wpa_supplicant and they never call `SSL_set_session` anywhere.)

indutny replied:
Hm... but version is still negotiated, even in EAP-FAST case, right? Couldn't it be that we got EAP master secret for higher version than what we just negotiated here? Or is it ok because it just sets `master_secret`, which is equally sized for all of TLS1.x?

indutny replied:
> (I've just checked wpa_supplicant and they never call SSL_set_session anywhere.)

I have no idea what you are talking about :)

davidben replied:
It doesn't look like RFC 4851 says anything about the PAC being somehow version-specific. I dunno, I suspect they didn't think about this very hard. Certainly the `tls_session_secret_cb` never fills an `ssl_version` value in.

>> (I've just checked wpa_supplicant and they never call SSL_set_session anywhere.)

> I have no idea what you are talking about :)

Normally, we get a useful `ssl_version` to check against because the previous `SSL_SESSION` had some value when it was established. Those come in via `SSL_set_session`. Since the EAP-FAST code assumes you never use it and `tls_session_secret_cb` doesn't fill in `ssl_version`, you're not going to get anything to check against.

(Or were you asking about why wpa_supplicant? This function was added for wpa_supplicant. They had some hacky patch for OpenSSL that eventually got upstreamed. I certainly hope no one else uses this code...)

indutny replied:
> It doesn't look like RFC 4851 says anything about the PAC being somehow version-specific. I dunno, I suspect they didn't think about this very hard. Certainly the tls_session_secret_cb never fills an ssl_version value in.

Looks like you are right about it, fixed.

> (Or were you asking about why wpa_supplicant? This function was added for wpa_supplicant. They had some hacky patch for OpenSSL that eventually got upstreamed. I certainly hope no one else uses this code...)

Yeah, I was asking about it. So you say that whole EAP thing is someone's hack for OpenSSL? ;)

indutny commented:
@davidben any idea what could be causing this spurious failure: https://travis-ci.org/openssl/openssl/builds/115612792 ?

davidben commented:
> @davidben any idea what could be causing this spurious failure: https://travis-ci.org/openssl/openssl/builds/115612792 ?

No idea. 25-test_verify.t is failing and I would think that doesn't even use the SSL code. A flake?

kroeckx commented:
That failure is a known problem as far as I know, were we have a test running on multiple docker instances at the same time trying to bind to the same port.  We didn't have the time yet to write something that avoids that problem.

kroeckx commented:
And it was 70-test_sslsessiontick.t that was failing, not 25-test_verify.t?

indutny commented:
@kroeckx is this a known problem too?

kroeckx commented:
Not sure what you mean with "this"?

indutny commented:
@kroeckx sorry for confusion, I meant `70-test_sslsessiontick.t`.

kroeckx commented:
So to be clear, 70-test_sslsessiontick.t and a few others are known to have this issue, 25-test_verify.t is not known to have it.

indutny commented:
Ok, good. Thank you!

Is there anything else that needs to be done/fixed before landing this?

kroeckx commented:
Not sure yet.  I still need to look at it carefully, and we'd need 2 people from the team to approve it.

davidben commented:
Not on the OpenSSL team, but the change lgtm. (Though I haven't looked at the test stuff since I'm not familiar with OpenSSL's testing framework.)

davidben commented:
I would suggest OpenSSL invest in some tests to assert resumption happens when it should (and doesn't when it shouldn't) on both client and server though. It seems both https://github.com/openssl/openssl/pull/852#issuecomment-195847997 and https://rt.openssl.org/Ticket/Display.html?id=4392 sneaked past review.

kroeckx commented:
I agree that we need more tests in various areas.  There are lots of things that can be improved, and I guess it's a matter of priorities.

indutny commented:
@kroeckx please let me know if I could do anything to help you guys with review. Thanks again!

kroeckx commented:
We're planning a release of the beta tomorrow.  This change will have to wait until after it.

indutny commented:
Kindly reminder about this issue (cc @kroeckx and @mattcaswell )

indutny commented:
I don't want to add pressure, but kindly reminding about it again. Hopefully it will help to move this issue to the top of your inbox :wink: 

_(Though, I understand how stressing reviewing security-related Pull Requests can be, so please forgive me!)_

mattcaswell commented on ssl/statem/statem_clnt.c:
```diff
@@ -821,7 +821,10 @@ int tls_construct_client_hello(SSL *s)
+    if ((sess == NULL) ||
+        /* Only destroy session which version is bigger than maximum supported.
+         */
+        ssl_version_cmp(s, sess->ssl_version, s->version) > 0 ||
```
I'm not sure this is quite right. What if we are deliberately using a version locked method and the session version is different? Or what if we are using a version flexible method, but the session version is disabled via SSL_OP_NO_*?

indutny replied:
I agree that this should handle `SSL_OP_NO*` too, but I don't really follow your previous statement. Sorry!

mattcaswell replied:
I'm talking about the version locked methods. So, for example, if someone has used TLSv1_1_client_method() (instead of the version flexible TLS_client_method()) but the session version is TLSv1.0. According to this logic, that would be ok, but actually it's doomed to fail.

indutny replied:
I think this scenario will actually work just fine. As far as I see it, when `TLSv1_1_client_method` - the `ssl_choose_client_version` will leave `s->version` untouched (and equal to `ctx->version`), and session version check below will fail, unless the session was created just for this connection. Does it make sense?

mattcaswell replied:
Which session version check below? You mean the one in `tls_process_server_hello`? That's too late - we're doomed to fail then - we've asked the server to resume with a session which we're not prepared to use.

indutny replied:
Oh right, that was different function. Sorry!  Should it just check for `any` methods here, then? And if method is not one of `any`, do exact version comparison (`!= 0`)

davidben replied:
It's more "doomed to be rejected" rather than "doomed to fail" (i.e. it's not a problem for the connection, just pointless to offer), but that's a good point. The check should actually be the same as `tls_process_server_hello` would do, but without the actual version switch.

(I should go fix that in BoringSSL too. Thanks!)

mattcaswell replied:
I think it is doomed to fail isn't it (as long as the server resumes the session)?? We are talking about a version locked method requesting a session for a different version; the server resuming that session and therefore using that different version, which the client will then abort the connection because it is not prepared to talk that version.

indutny replied:
I believe it should be fixed now, PTAL

davidben replied:
@mattcaswell Servers should be doing version negotiation before session resumption and won't get into that situation. Any server that does this in the wrong order already cannot interoperate with other implementations. No client implementation (not even OpenSSL, due to this version locking) accepts a session/version mismatch, and most clients are not version-locked, so they will freely offer a TLS 1.1 session on a TLS 1.2 ClientHello.

Plus the particulars of session resumption are version-specific (TLS 1.3 is due to change it very significantly), so doing it in the wrong order isn't meaningful.

mattcaswell commented on ssl/statem/statem_clnt.c:
```diff
@@ -1130,12 +1133,24 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL *s, PACKET *pkt)
+    if (s->version != s->session->ssl_version) {
+        al = SSL_AD_PROTOCOL_VERSION;
+
+        /* XXX(indutny): introduce new error code */
```
This comment needs resolution

indutny replied:
I have a question about this: it looks like the error codes are declared without any order. Should I just pick any free error code, or is there some other convention here for introducing a new ones?

mattcaswell replied:
Give it the name you want to give it in this file. Then run `make errors`. The error codes will be automatically assigned.

indutny replied:
Thanks, will do!

indutny replied:
A question: it seems that `make error` has inadvertently modified `SSL_F_CT_MOVE_SCTS` and some other errors to make their string description lower-case. Should I commit these changes, or just revert them manually?

mattcaswell replied:
Probably someone forgot to run `make errors` previously. Just commit the changes relevant to you.

indutny replied:
Ack, pushed.

mattcaswell commented on test/ssltest.c:
```diff
@@ -1829,24 +1876,53 @@ int main(int argc, char *argv[])
-        goto end;
-    if (!set_protocol_version(client_max_proto, c_ssl, SSL_CTRL_SET_MAX_PROTO_VERSION))
-        goto end;
+                sess = dup_session(SSL_get_session(c_ssl));
```
Not sure I understand why dup_session is necessary. Can't we use SSL_SESSION_up_ref() here?

indutny replied:
I want to reset method and every internal field in session too, but will give a try to `SSL_SESSION_up_ref()`. It is very likely that it will work just fine too.

mattcaswell commented on ssl/statem/statem_lib.c:
```diff
@@ -704,6 +704,19 @@ static int version_cmp(const SSL *s, int a, int b)
     return DTLS_VERSION_LT(a, b) ? -1 : 1;
 }
 
+int ssl_strict_version_check(const SSL *s, int version)
```
I think this still isn't quite right. Compare this with the checks in `ssl_method_error`.

indutny replied:
May I ask you to elaborate on this?

It really looks like that `min`/`max` versions will be handled later in `ssl_choose_client_version` during `tls_process_server_hello`. Is there any point in doing them during `ssl_strict_version_check`?

mattcaswell replied:
We don't want to offer a session for a version we are not prepared to accept from the server. The ssl_method_error function also does the check for SSL_OP_NO_*.

indutny replied:
Do you mean `(s->options & method->mask) != 0` ?

mattcaswell replied:
Yes

indutny replied:
Ack, please check.

kroeckx replied:
Can you please document this function?

indutny replied:
Does this looks good?

```
/* Check that the specified `version` is supported by `SSL *` instance */
```

kroeckx commented on ssl/statem/statem_lib.c:
```diff
@@ -704,6 +704,44 @@ static int version_cmp(const SSL *s, int a, int b)
+    /* Version should not be disabled by SSL_set_options */
+    if ((s->options & mask) != 0)
+        return 0;
+
```
This seems to at least fail to check SSL_METHOD_NO_FIPS and SSL_METHOD_NO_SUITEB.

kroeckx replied:
Is there a reason this can't just use ssl_method_error()?

indutny replied:
The reason is that we don't have a `method` matching session version during this call. I guess I could turn this `switch` into one that will select `method` instead of `mask`, and then use `ssl_method_error` instead? Does it make sense?

kroeckx replied:
I thought about just going over (d)tls_version_table and checking that the version is enabled like ssl_get_client_min_max_version() does, and stop when the version from the table gets lower than the one from the session.

indutny replied:
Pushed an update with docs, PTAL. Thank you!

indutny commented:
Ping ;)

richsalz commented:
Various folks are away, often for easter around now.

indutny commented:
Aah, no worries. Have a happy Easter then! Thank you @richsalz

kroeckx commented:
Or some have a daytime job that's unrelated to OpenSSL ...

indutny commented:
Ok ok! I regret pinging you guys now! :dash:

kroeckx commented on ssl/statem/statem_lib.c:
```diff
@@ -788,6 +788,52 @@ static int ssl_method_error(const SSL *s, const SSL_METHOD *method)
+           * is an exact match. In any case - our search is over now
+           */
+          return cmp == 0;
+        }
```
I was expecting something like:

    for (vent = table; vent->version != 0 && version_cmp(s, version, vent->version) >= 0; vent++) {
        if (vent->cmeth != NULL && ssl_method_error(s, vent->cmeth()) == 0)
            return 1;
    }
    return 0;

Note that I use cmeth(), not smeth().

indutny replied:
Ack.

indutny replied:
On other hand... I don't think that it is strict enough. What I am trying to figure out is if there is an enabled method that allow switching to the specified version. Will push an update with a mix of both of our versions.

kroeckx commented:
It's my understanding that this patch tries to solve the problem that even when both sides support a higher protocol version that it might still use the older from the session.  If that's not the case, could you please clarify things?

If I properly understand things, this change will actually break things for clients that try to talk to older openssl servers, before commit 03d14f58873470407de6120218b7e69fefd8b58f.  Maybe other implementations also have issues like this.

So it becomes a question about which behaviour is worse, break a connection, or stay with an older protocol (for a while)?

There are references to bugs people have with amazon upgrading and downgrading the version, and than in ssl3_get_record() then complains that it's expecting a different version.  From what I understand this happens on downgrade, not on upgrade.

I've read that we've fixed something in the master branch during the rewrite.  Is that the problem that we're expecting some version but get a different one?

So I have 2 concerns:
- We're mixing up things and not actually fixing the problem people actually run in to
- We might break something with this

kroeckx commented:
So looking at what I get back from amazon, it seems to be that it just sends back a fatal alert 70 (protocol version) in case I try to resume TLS 1.2, so there seems little we can do about that.

indutny commented:
> It's my understanding that this patch tries to solve the problem that even when both sides support a higher protocol version that it might still use the older from the session. If that's not the case, could you please clarify things?

While you are definitely right in this, the initial intention of the patch was to solve problem with a pool of servers, when someone of them have lower version than the most of the others (AWS case). If client would cache SSL sessions (as we do in node.js), it will eventually try to establish a connection with one of the lower-version servers, and this connection will fail because of the client-side version mismatch error.

Thus, this PR attempts to make SSL session caching useful for AWS, Ebay API, and possibly other big pools of TLS servers under the same DNS hostname.

> If I properly understand things, this change will actually break things for clients that try to talk to older openssl servers, before commit 03d14f5. Maybe other implementations also have issues like this.

I'm not sure how it will break servers, honestly saying... It should be safe to assume that the modern-version servers should not share session cache with the lower-version servers. Of course, if they accept the session id, client will assume that they can speak higher protocol version than what they actually can.

Anyway this problem can happen now too. Imagine a client speaking (to the same) pool of servers, they can also cache a session from the old server, and attempt to send it to a newer server (which may have that older protocol disabled). Regardless, of acceptance of this PR, this scenario is very likely going to lead to some failure.

> So it becomes a question about which behaviour is worse, break a connection, or stay with an older protocol (for a while)?

I home my previous comment explained, that it is not better or worse with regards to that problem, but it (at least) solves another.

> There are references to bugs people have with amazon upgrading and downgrading the version, and than in ssl3_get_record() then complains that it's expecting a different version. From what I understand this happens on downgrade, not on upgrade.

Usually yes, because higher-version servers support old protocols too.

> I've read that we've fixed something in the master branch during the rewrite. Is that the problem that we're expecting some version but get a different one?

Passing SSL_SESSION to client before this PR switched the client's method from ANY to particular TLS1.0/1.1/1.2 method, depending on the version stored in session. Of course this leads to version mismatch error in case if TLS1.0 server will reply to the client with TLS1.2 method.

>So I have 2 concerns:
> We're mixing up things and not actually fixing the problem people actually run in to

I don't think this is the case, and we are fixing the problem that real people really run into.

> We might break something with this

This is absolutely possible, but with your, @mattcaswell, and @davidben review the risk should be minimal possible.

> So looking at what I get back from amazon, it seems to be that it just sends back a fatal alert 70 (protocol version) in case I try to resume TLS 1.2, so there seems little we can do about that.

It is very interesting to learn about. In my tests, applying this patch fixed the problem completely when talking to Ebay API and AWS...

indutny commented:
@kroeckx pushed fixes.

kroeckx commented:
Do you happen to have any network captures of trying to resume a TLS 1.2 connection talking to a 1.0 server?

kroeckx commented:
I was getting confused because of how wireshark showed things.  It's showing the ServerHello after we've send back the alert message to it, because we didn't receive the whole ServerHello message in a single packet.

indutny commented:
Here is one:

[tls.pcap.zip](https://github.com/openssl/openssl/files/188736/tls.pcap.zip)

And here is a node.js script, that I used to generate it:

https://gist.github.com/indutny/56b0c9dcae27ff21b6f9

(Note, no particular reason for `AES128-SHA`, just a habit of disabling PFS before capturing tcpdump data).

kroeckx commented:
So if I understand things correctly, this patch is fixing 2 issues:
- Don't think we should get back the version from the session, it might not get resumed
- Announce a higher version that's supported even when trying to resume a session that has an older version, so that we don't get stuck with the old version.

I think we clearly need to fix the first issue, but I think change for the 2nd might cause interoperability issues, so I prefer to first try to fix the first issue.

indutny commented:
I absolutely agree with you with regards to the first issue of two. However, I believe that second is as necessary as the first one.

It seems to me that second issue with very important, because not fixing it will lead to the version (and security) downgrade for a whole pool after running app with a session cache for some time. See, as we establish more and more connections with the pool - we are inevitably going to meet TLS1.0 nodes, and will store the TLS1.0 session in our cache. Moreover, in your scenario we are not going to advertise TLS1.2 anymore, so practically all connections will eventually degrade to TLS1.0, even considering that TLS1.2 is supported by both server and the client.

I also have a feeling that not advertising higher version may lead to some (possibly unobvious) version degradation attacks, but I can't really think of a particular scenario though.

kroeckx commented:
The reason for not advertising the higher version is that the server might do the wrong thing in case the versions are different.  And openssl did the wrong thing before commit 03d14f58873470407de6120218b7e69fefd8b58f.  It resumed the session when the versions where different, but with the new version, not the one from the session. I'm guessing almost nobody is actually using that fix.  I assume the reason this problem was found and fixed is because boringssl started send the higher version.

kroeckx commented:
So my understanding of what I think should be fixed is that when we want to resume a session (as client) in SSL_set_session() we now switch to a fixed version method while we shouldn't because we don't know if we're going to be resumed or not.  Not switching to that method results in us now announcing the highest supported version, but I think we should instead use the one from the session.

indutny commented:
I think the reasons for the fix in BoringSSL may be different from what you
supposed. Cc @davidben

While I see your point and reasoning, it doesn't look like benefits from
such limited fix would overweight its negative sides (see pool version
downgrade scenario from my previous comments).

Perhaps @davidben may want to share some thoughts on this side as well.

On Thursday, March 24, 2016, Kurt Roeckx <notifications@github.com> wrote:

> So my understanding of what I think should be fixed is that when we want
> to resume a session (as client) in SSL_set_session() we now switch to a
> fixed version method while we shouldn't because we don't know if we're
> going to be resumed or not. Not switching to that method results in us now
> announcing the highest supported version, but I think we should instead use
> the one from the session.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/852#issuecomment-201065248>
>

davidben commented:
It's not quite stay at old version without the fix vs break with it. There are many situations where OpenSSL's current behavior breaks things. It's messy and interacts with many other behaviors. (At one point I had to make a chart of 4 or 5 other quirks that combine.) For simple cases, see https://crbug.com/417134 what we ran into in Chrome. This [Twitter thread](https://twitter.com/pzb/status/531677249909821440) is another example of what can go wrong.

For the pre 03d14f58873470407de6120218b7e69fefd8b58f bug: yes, that is a risk. But note it's harder to trigger than the above risks you have if you don't take this patch. You a multi-version deployment *that shares a session cache* (or session ticket key). This takes more expertise than simply having a heterogeneous mix of machines with your cert.

Still, when I was first doing these changes in BoringSSL, I too was worried about this and initially did not enforce a match between `session->ssl_version` and `s->version`. If OpenSSL is worried, you can do something similar. But I do not think you should do this.

When I implemented the lax version, I shipped metrics in Chrome to see if we could enforce things. It turned out well. BoringSSL (and thus Chrome) has had the behavior in this patch for about a year now with no sign of issues. [As has Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1134548). To my knowledge, they have not run into issues either.

> I've read that we've fixed something in the master branch during the rewrite. Is that the problem that we're expecting some version but get a different one?

I am not aware of anything around this area that the rewrite fixed. It seems to have inherited all of 1.0.2's behavior.

indutny commented:
Thank you, David! I couldn't say it better.

On Thursday, March 24, 2016, David Benjamin <notifications@github.com>
wrote:

> It's not quite stay at old version without the fix vs break with it. There
> are many situations where OpenSSL's current behavior breaks things. It's
> messy and interacts with many other behaviors. (At one point I had to make
> a chart of 4 or 5 other quirks that combine.) For simple cases, see
> https://crbug.com/417134 what we ran into in Chrome. This Twitter thread
> <https://twitter.com/pzb/status/531677249909821440> is another example of
> what can go wrong.
>
> For the pre 03d14f5
> <https://github.com/openssl/openssl/commit/03d14f58873470407de6120218b7e69fefd8b58f>
> bug: yes, that is a risk. But note it's harder to trigger than the above
> risks you have if you don't take this patch. You a multi-version deployment *that
> shares a session cache* (or session ticket key). This takes more
> expertise than simply having a heterogeneous mix of machines with your cert.
>
> Still, when I was first doing these changes in BoringSSL, I too was
> worried about this and initially did not enforce a match between
> session->ssl_version and s->version. If OpenSSL is worried, you can do
> something similar. But I do not think you should do this.
>
> When I implemented the lax version, I shipped metrics in Chrome to see if
> we could enforce things. It turned out well. BoringSSL (and thus Chrome)
> has had the behavior in this patch for about a year now with no sign of
> issues. As has Firefox
> <https://bugzilla.mozilla.org/show_bug.cgi?id=1134548>. To my knowledge,
> they have not run into issues either.
>
> I've read that we've fixed something in the master branch during the
> rewrite. Is that the problem that we're expecting some version but get a
> different one?
>
> I am not aware of anything around this area that the rewrite fixed. It
> seems to have inherited all of 1.0.2's behavior.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/852#issuecomment-201069470>
>

davidben commented:
Oh, we actually crossed mid-stream, but I guess I ended up answering your summons after all. :-)

davidben commented:
I found my chart of behaviors. If my notes are right, Safari and IE11 also have this behavior...
...that said, both of them (and Firefox) included the IP address in the session cache key[*], which lowers the risk somewhat.
...that said, Chrome does not key by IP address and has been fine.
...that said, the browser version fallback can work around the bug.
...that said, Firefox has since removed the fallback. I've mostly removed it from Chrome. The next major release should finish the job.

It's a mess. :-) Due to all the weird interactions, it's hard to translate one implementation's experience onto another, but I believe you shouldn't be worried about it.

[*] OpenSSL leaves this to the consumer, so this depends.

kroeckx commented:
In the Mozilla bug it says you saw a 0.06% increase in error rate because of this change, and 0.02% after splitting of some cache.

davidben commented:
Yeah, this is what I mean by subtle interactions. :-) If you are a browser without a version fallback, do not shard the session cache between fallback legs, *and* do not have one of several quirks that results in a sticky fallback, the OpenSSL server-side bug will trigger on spurious fallbacks.

To be honest, I'm not sure what happened to that 0.02% predicted mass, but when we started actually enforcing, it never appeared anywhere. At the time, I decided I was okay letting the version fallback weather any instances. Today, session version mismatch is a tiny fraction of connections that needed the fallback, which is itself [a tiny fraction of](https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/yz1lU9YTeys/yCsK50I3CQAJ) overall connections.

kroeckx commented:
The current version doesn't seem to try to resume the session in case the session has an older protocol.

indutny commented:
@kroeckx you mean server or client side? Client side definitely does if its method allows it.

kroeckx commented on ssl/statem/statem_lib.c:
```diff
@@ -788,6 +788,44 @@ static int ssl_method_error(const SSL *s, const SSL_METHOD *method)
+    }
+
+    for (vent = table;
+         vent->version != 0 && version_cmp(s, version, vent->version) >= 0;
```
That should be `<=` instead of `>=`

indutny replied:
Ack, fixed!

kroeckx commented on ssl/statem/statem_lib.c:
```diff
@@ -788,6 +788,44 @@ static int ssl_method_error(const SSL *s, const SSL_METHOD *method)
+ *
+ * Returns 1 on success, otherwise 0
+ */
+int ssl_strict_version_check(const SSL *s, int version)
```
Could you maybe rename this to ssl_version_supported()?

And change the comment to "Returns 1 when supported"

indutny replied:
Ack, thank you!

kroeckx commented on ssl/statem/statem_clnt.c:
```diff
@@ -821,7 +821,11 @@ int tls_construct_client_hello(SSL *s)
+        /* Only destroy session which version is bigger than maximum supported,
+         * or session which version is not supported by the current ssl->method
+         */
+        !ssl_strict_version_check(s, sess->ssl_version) ||
```
I think this comment should just say that it's not supported.

indutny replied:
Do we need comment here at all? The name seems to be much more descriptive now.

indutny replied:
Preemptively removed comment, ack.

kroeckx commented:
So there is 1 thing I don't fully understand yet, and that's that you might keep stuck at an older version even when both sides support a newer.

As far as I can see this only happens when a client tries to resume an expired session.  OpenSSL just takes whatever is passed to SSL_set_session() and doesn't check that it's expired or not and leaves it up to the application to remove the expired sessions.

indutny commented:
@kroeckx this is definitely possible. I suppose that application will evict sessions itself, or just ignore sessions with lower versions when connecting to a pool where majority of nodes support higher version. Anyway, it looks like an application logic to me, not something that we can fix here.

davidben commented:
Yeah, it depends on how frequently the client application evicts expired sessions, if it does at all. OpenSSL leaves client-side session management to the consumer, so it varies by consumer.

(It does have to specifically be client-side session expiration though. Whether the server's expired the session doesn't figure into whether the client decides to offer it.)

kroeckx commented:
Patch is applied in master.

indutny commented:
Thank you Kurt, Matt, David!
